x-pack/plugins/fleet/server/services/metrics/fetch_agent_metrics.ts

```diff
@@ -0,0 +1,260 @@
+                },
+              },
+            ],
+          },
+        },
+        size: 0,
+        aggs: {
+          upgrade_details: {
+            terms: { field: 'upgrade_details.state' },
+          },
+        },
+      },
+      { signal: abortController.signal }
+    );
+    ((response?.aggregations?.upgrade_details as any).buckets ?? []).forEach((bucket: any) => {
+      switch (bucket.key) {
+        case 'UPG_REQUESTED':
+          upgradingSteps.requested = bucket.doc_count;
+          break;
+        case 'UPG_SCHEDULED':
+          upgradingSteps.scheduled = bucket.doc_count;
+          break;
+        case 'UPG_DOWNLOADING':
+          upgradingSteps.downloading = bucket.doc_count;
+          break;
+        case 'UPG_EXTRACTING':
+          upgradingSteps.extracting = bucket.doc_count;
+          break;
+        case 'UPG_REPLACING':
+          upgradingSteps.replacing = bucket.doc_count;
+          break;
+        case 'UPG_RESTARTING':
+          upgradingSteps.restarting = bucket.doc_count;
+          break;
+        case 'UPG_WATCHING':
+          upgradingSteps.watching = bucket.doc_count;
+          break;
+        case 'UPG_ROLLBACK':
+          upgradingSteps.rollback = bucket.doc_count;
+          break;
+        case 'UPG_FAILED':
+          upgradingSteps.failed = bucket.doc_count;
+          break;
+        default:
+          break;
+      }
+    });
+    return upgradingSteps;
+  } catch (error) {
+    if (error.statusCode === 404) {
+      appContextService.getLogger().debug('Index .fleet-agents does not exist yet.');
+    } else {
+      throw error;
+    }
+    return upgradingSteps;
+  }
+};
+
+export const getUnhealthyReason = async (
+  esClient: ElasticsearchClient,
+  abortController: AbortController
+): Promise<UnhealthyReason> => {
+  const unhealthyReason = {
+    input: 0,
+    output: 0,
+    other: 0,
+  };
+  try {
+    const response = await esClient.search(
+      {
+        index: AGENTS_INDEX,
+        query: {
+          bool: {
+            filter: [
+              {
+                term: {
+                  active: 'true',
+                },
+              },
+              {
+                terms: {
+                  last_checkin_status: ['error', 'degraded'],
+                },
+              },
+            ],
+          },
+        },
+        size: 10000,
+        _source: ['components'],
+      },
+      { signal: abortController.signal }
+    );
+
+    response.hits.hits.forEach((hit: any) => {
+      // counting agent as other unhealthy reason if it doesn't have a component unit unhealthy
+      if (!hit._source.components || hit._source.components.length === 0) {
+        unhealthyReason.other++;
+        return;
+      }
+      // considering component unhealthy if not healthy
+      const hasUnhealthyUnit = (type: string) => {
+        const unhealthyComponent = hit._source.components.find(
+          (comp: any) => comp.status !== 'HEALTHY'
```

Comment:
Assuming statuses other than healthy mean unhealthy, I am a bit doubtful on things like STARTING, though it's an issue if the agent is stuck in that.
https://github.com/elastic/kibana/blob/b641a6998540a3f8f565d32fa3cac2b657ffd4c5/x-pack/plugins/fleet/common/constants/agent.ts#L29-L36